File: January 2025/08.c
#include <stdio.h>
#include <stdlib.h>

// Function to compare two integers for qsort
int compare(const void *a, const void *b) {
    return (*(int *)a - *(int *)b);
}

// Function to count the number of possible triangles
int countTriangles(int arr[], int n) {
    // Sort the array
    qsort(arr, n, sizeof(int), compare);
    int count = 0;

    // Iterate through the array using three pointers
    for (int i = 0; i < n - 2; i++) {
        int k = i + 2;
        for (int j = i + 1; j < n - 1; j++) {
            while (k < n && arr[i] + arr[j] > arr[k]) {
                k++;
            }
            // Count the number of possible triangles with arr[i] and arr[j]
            if (k > j) {
                count += k - j - 1;
            }
        }
    }

    return count;
}

File: January 2025/08.py
class Solution:
    # Function to count the number of possible triangles.
    def countTriangles(self, arr):
        # Sort the array.
        arr.sort()
        n = len(arr)
        count = 0
        
        # Iterate through the array using three pointers.
        for i in range(n - 2):
            k = i + 2
            for j in range(i + 1, n - 1):
                while k < n and arr[i] + arr[j] > arr[k]:
                    k += 1
                # Count the number of possible triangles with arr[i] and arr[j]
                if k > j:
                    count += k - j - 1
        
        return count
